### PR TITLE
✨ Server-side sessions

### LINKED ISSUE #207: Session mgmt
### Business Decisions
1. Should existing user sessions be preserved?
   - If no, all users will be logged out, and have to log in again (once)
   - If yes, users will be able to continue using the apps and will be gradually migrated onto the new session system
   - If yes, this will add ~30% to the development effort
2. Should old sessions be monitored?
   - Allows us to check how long until old system can be retired
   - Adds to development effort
   - Potentially limited use, if transition period is on order of 1-2 weeks.

### Technical Decisions
1. If existing sessions are preserved, two options:
   1. Create new scheme/strategy which handles both JWT and session IDs
   2. Create new scheme/strategy which handles session IDs, add this strategy to all relevant API routes
2. How to reliably test before deployment?

#### Migration under Option 1
The `validate` method:
  - checks for presence of session cookie.
    - If present and valid, continue.
    - If present and invalid, 401.
    - If not present, 401.
  - checks for presence of JWT.
    - If present and valid, sets a session cookie.
    - If present and expired, 401.
    - If present and invalid, 401.
    - If not present, 401.
  - 401

Login route is changed to:
  - Authenticate against the new strategy in `try` mode
  - If authenticated with JWT:
     - Set session ID
     - 200
  - If authenticated with session ID:
     - 200
  - Else evaluate payload:
     - If valid, create new session, set session ID cookie
     - Else 401

Only users presenting a valid and un-expired JWT are given a new session. If they log-out and -in, they are switched to the new system. If they don't, they will be switched once their existing JWT expires (max 1 week). Then after all the JWT are expired, change the `validate` method to remove the JWT branch.

### Pros
- Code changes localised to one area
- Unlikely to lead to missed endpoints

### Cons
- Logic is more complex (temporarily)
- Configuration is more complex (temporarily)

#### Migration under Option 2
The `validate` method is identical to the first branch of Option 1. The new strategy should be positioned as first in the list of strategies for all API routes.

Login route is changed to:
  - Authenticate against the new strategy in `try` mode
  - If request has JWT:
     - Set session ID
     - 200
  - If authenticated with session ID:
     - 200
  - Else evaluate payload:
     - If valid, create new session, set session ID cookie
     - Else 401

Users with a valid and un-expired JWT are allowed to continue using the app. If they log-out and -in, they are switched to the new system. If they don't, they will be switched once their existing JWT expires (max 1 week).

### Pros
- Logic is simpler
- Configuration is simpler?

### Cons
- Code changes spread throughout the code base
- Could lead to missed endpoints

#### Testing Before Deployment
If sessions are not preserved, this is simple.

If sessions are preserved:
- must be authenticated with valid JWT
- must be authenticated with valid session ID
- login with valid JWT must return session ID (regardless of payload)
- login without valid JWT must return session ID
- expired JWT must give 401

If sessions are not preserved:
- must NOT be authenticated with valid JWT
- must be authenticated with valid session ID
- login with valid JWT must return session ID (if payload is correct)
- login with valid JWT must return 401 (if payload is incorrect)
- login without JWT must return session ID
- expired JWT must give 401

## Tasks
Note that not all these tasks are necessary, depending on the result of the business and technical decisions.

1. Setup redis config and connection (1 day)
1. Create new schema (1.5 day)
1. Create new strategy (0.5 days - 1.5 days)
1. Update routes (0 - 1 days)
1. Changes to login handler (1 day)
1. Review (1 day)

Estimate: 5 - 7.5 days

### PR DESCRIPTION
**Pull request by eliascodes**
fixes #207 

~80% of the diff is renaming and config changes.

Manually tested with visitor, volunteer and dashboard apps.

### Changes
- Use `@hapi/yar` and `@hapi/catbox-redis` to setup server-side session cache
- Create `session_cookie` authentication schema to validate requests using the session cache
- Re-write the `standard` authentication strategy to use the `session_cookie` schema and rely on `yar`
- Add config for the new auth system and the session cache
- Add a bit of documentation
- Update tests
- Remove pointless `src/routes` file
- Remove `hapi-auth-jwt2` and `jsonwebtoken`

### Testing Requirements

- [x] **Visitor app**

- Sign in
- Register visit
- Go to admin area
- Visit an admin page
- Sign out
- Try to (unsuccessfully) navigate directly to admin area (by typing in URL)

- [x] **Volunteer app**
- Sign in
- Create log
- Update user info
- Sign out

- [x] **Dashboard**
- Sign in
- Click one of the tiles
- Check data loads
- Sign out

### Release Requirements
- Thursday 8th August

### Manual Deployment
- [ ] **API**
- [ ] **Visitor app**
- [ ] **Dashboard**

### CODE REVIEW

File: api/config/config.defaults.ts [hunk 1]
```diff
@@ -89,7 +89,7 @@ const config: DeepPartial<Config> = {
         options: {
           name: 'twine-api-session',
           cookieOptions: {
-            password: 'pefwj48208tfh29pd9cm10cprmhx94pfry28ctnc2mkrx0dufpog3',
```
Comment (astroash) [inline api/config/config.defaults.ts]: Just checking this does not need to be removed from git history

File: api/config/__tests__/index.test.ts [hunk 2]
```diff
@@ -134,18 +144,13 @@ describe('Config', () => {
...
-              isSecure: true,
-              isHttpOnly: true,
-              isSameSite: 'Lax',
-              path: '/',
+        schema: {
+          session_cookie: {
+            options: {
+              name: 'wpppp',
+              cookieOptions: {
+                password: 'rt49089y5h3j9u0894thugirbv90u8r9g7tu4fevu9rguib2fwe',
```
Comment (astroash) [inline api/config/__tests__/index.test.ts]: 👀

File: api/config/config.defaults.ts (hunk 1 above)
Comment (eliascodes) [reply api/config/config.defaults.ts]: I think it's fine, it was just me mashing the keyboard for a temp password to use while developing. 

I've not used the same password for the environment variable in the `.env` file locally and wont use the same one in the env vars on the hosted environments.

**Comment by astroash** [comment]
@eliascodes what's the release status of this PR?

**Comment by eliascodes** [comment]
it was not part of the recent release, it hasn't been merged

**Comment by astroash** [comment]
Sorry, that wasn't clear. Are we okay to release this when ever or is there going to be comms about everyone being signed out of the app.

**Comment by eliascodes** [comment]
> Is this being QA'd?

Don't think so given Amanda's last day is Monday/Tuesday

> Are we okay to release this when ever or is there going to be comms about everyone being signed out of the app.

We can ask about that on Monday, but I imagine there won't be any comms.

**Comment by astroash** [comment]
Can you please add what you consider needs manual testing in the format that we used to use:
- bullet points in body of PR
- apps to test
- flows that will be affected

**Comment by astroash** [comment]
@eliascodes can you update this PR in line with https://github.com/TwinePlatform/twine-monolith/issues/238#issuecomment-517657111

**Comment by eliascodes** [comment]
Discovered minor security issue explained in #258. IMO very low risk right now.

**Comment by eliascodes** [comment]
Also LMK whether you'd prefer to address the `@types/ramda`/yarn workspaces thing before this is merged, or whether you're OK with the hack staying temporarily. I'm easy.

**Comment by astroash** [comment]
> Also LMK whether you'd prefer to address the @types/ramda/yarn workspaces thing before this is merged, or whether you're OK with the hack staying temporarily. 

Let's try and fix it today, but release as is if it's not done by the afternoon.

File: api/docs/configuration.md [hunk 3]
```diff
@@ -16,6 +16,12 @@ DATABASE_URL=...
 DATABASE_URL_DEVELOPMENT=...
 DATABASE_URL_TESTING=...
 
+# Redis connection strings for redis server in various environments
+# REDIS_URL is used in production, the others are suffixed with their environments
+REDIS_URL=redis://h:pad3d01b1b7a5335ba4b783eee45025181dc379ab5df46d57099b37474527d104@ec2-3-223-92-209.compute-1.amazonaws.com:27339
```
Comment (astroash) [inline api/docs/configuration.md]: 👀

**Pull request merged**